### PR TITLE
Add governence token staking

### PR DESCRIPTION
**Pull request by btree1970**
Add staking functionality to the LibToken. Staking is represented as a ERC20 token

### CODE REVIEW

File: test/PoolFactory.js [hunk 1]
```diff
@@ -2,42 +2,48 @@ const { expect } = require("chai");
...
-    [address1, admin1, admin2] = await ethers.getSigners();
+    poolFactoryInstance = await poolFactoryContract.deploy(skLibToken.address);
 
-    poolFactoryToken = await poolFactoryContract.deploy();
-    await poolFactoryToken.deployed();
+    await poolFactoryInstance.deployed();
   });
 
-  
-  describe("PoolFactory dynamics", function () {
```
Comment (ksgebrekirstos98) [inline test/PoolFactory.js:17]: Why did you remove this? We need it to test the PoolFactory functions

File: test/stkLibToken.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+describe("stkToken contract", function () {
+
+  let alice;
+  let bob;
+  beforeEach(async function () {
+
+    [alice, bob] = await ethers.getSigners();
+
+    libFactory = await ethers.getContractFactory("LibToken");
```
Comment (ksgebrekirstos98) [inline test/stkLibToken.js:16]: add ```const``` everywhere it applies

**Review by ksgebrekirstos98 — CHANGES_REQUESTED** [review]

File: test/PoolFactory.js [hunk 3]
```diff
@@ -2,42 +2,48 @@ const { expect } = require("chai");
...
+
+      // Stake LibToken
+      await skLibToken.stake(amount)
+      expect(await skLibToken.balanceOf(alice.address)).to.equal(amount)
+      expect(await LibToken.balanceOf(alice.address)).to.equal(0)
+  }
+
+  it("Insufficient balance to create pool", async function () {
+      await StkToken(10000000000) //Insufficient balance
+      await expect( poolFactoryInstance.createPool("Whoopty", [bob.address, alice.address])).to.be.revertedWith("Insufficient staked balance to create pool")
```
Comment (ksgebrekirstos98) [inline test/PoolFactory.js:40]:
```suggestion
      await expect( poolFactoryInstance.createPool("TestPool", [bob.address, alice.address])).to.be.revertedWith("Insufficient staked balance to create pool")
```

File: test/PoolFactory.js [hunk 4]
```diff
@@ -2,42 +2,48 @@ const { expect } = require("chai");
...
+  });
+
+  it("Sufficient balance to create pool", async function () {
+      await StkToken(ethers.BigNumber.from("1000000000000000000000000000")) //sufficient balance
+      await expect( poolFactoryInstance.createPool("Whoopty", [bob.address, alice.address])).to.not.be.reverted
```
Comment (ksgebrekirstos98) [inline test/PoolFactory.js:45]:
```suggestion
      await expect( poolFactoryInstance.createPool("TestPool", [bob.address, alice.address])).to.not.be.reverted
```

File: test/PoolFactory.js (hunk 1 above)
Comment (btree1970) [reply test/PoolFactory.js:17]: fixed

File: contracts/governence/StkLibToken.sol [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    function unstake(uint256 amount) external nonReentrant {
+        uint256 balance = balanceOf(msg.sender);
+        require(balance >= amount, "Insufficent balanece to unstake");
+        _unstake(amount);
+    }
+
+    function _stake(uint256 amount) internal  {
+        require(amount > 0, "StakeLibToken: Cannot stake 0");
```
[the remark this answers is not in the document]
Comment (btree1970) [reply contracts/governence/StkLibToken.sol:40]: fixed

File: contracts/governence/StkLibToken.sol [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    // Address of LibToken contract
+    address public LibToken;
+
+
+    event Stake(address indexed staker, uint256 amount);
+    event Unstake(address indexed staker, uint256 burntAmount);
+
+    constructor(
+        address libToken,
+        string memory name,
```
[the remark this answers is not in the document]
Comment (btree1970) [reply contracts/governence/StkLibToken.sol:21]: to identify the contract

File: contracts/PoolFactory.sol [hunk 7]
```diff
@@ -13,13 +17,25 @@ contract PoolFactory {
      * Array to store all the created Pools
      */
     Pool[] public pools;
+    
+    address private StkToken;
+
+    uint256 private minStakeAmount = 1000000000000000000000;
```
[the remark this answers is not in the document]
Comment (btree1970) [reply contracts/PoolFactory.sol:23]: I just choose a random number feel free to fix it after to an appropriate amount

File: contracts/governence/StkLibToken.sol [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        emit Stake(msg.sender, amount);
+    }
+
+    function _unstake(uint256 amount) internal {
+        // Burn the stk tokens
+        _burn(msg.sender, amount);
+        
+       // Give user back their staked token
+        IERC20(LibToken).transfer(msg.sender, amount);
+        emit Unstake(msg.sender, amount);
```
Comment (tgetachew18) [inline contracts/governence/StkLibToken.sol:51]: Doesn't `transfer` already [emit](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/IERC20.sol#L24)? Is this for convenience?

File: contracts/governence/StkLibToken.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (ksgebrekirstos98) [reply contracts/governence/StkLibToken.sol:21]: Yeah, but we're not assigning them to any state variables.

File: test/PoolFactory.js (hunk 1 above)
Comment (ksgebrekirstos98) [reply test/PoolFactory.js:17]: Still looks deleted to me

**Review by ksgebrekirstos98 — APPROVED** [review]

File: test/PoolFactory.js (hunk 1 above)
Comment (btree1970) [reply test/PoolFactory.js:17]: @kaleb I have integrated it back, I just added the test into a different testing that is alreay added

File: contracts/governence/StkLibToken.sol (hunk 8 above)
Comment (btree1970) [reply contracts/governence/StkLibToken.sol:51]: hmm, this emit is only for staking, You can trasnfer without staking

File: test/PoolFactory.js (hunk 1 above)
Comment (ksgebrekirstos98) [reply test/PoolFactory.js:17]: Ohh, sounds good
